cs4660/graph/graph.py

```diff
@@ -0,0 +1,380 @@
+    - returns true if the edge is added and false if
+     the edge already existed
+* remove_edge
+
+    - remove an edge from its internal data structure
+    - returns true if the edge is removed and false if the edge does not exist
+"""
+
+
+
+
+from io import open
+from operator import itemgetter
+
+def construct_graph_from_file(graph, file_path):
+    """
+    TODO: read content from file_path, then add nodes and edges to graph object
+
+    note that grpah object will be either of AdjacencyList, AdjacencyMatrix or ObjectOriented
+
+    In example, you will need to do something similar to following:
+
+    1. add number of nodes to graph first (first line)
+    2. for each following line (from second line to last line), add them as edge to graph
+    3. return the graph
+    """
+    f = open(file_path, encoding='utf-8')
+    line = f.readline()
+    for i in range(int(line)):
+        graph.add_node(Node(i))
+    text = f.read()
+    lines = text.split('\n')
+    for line in lines:
+        if len(line) > 0:
+            values = line.split(':')
+            edge = Edge( Node( int( values[0] ) ), Node( int( values[1] ) ), int( values[2] ) )
+            graph.add_edge(edge)
+    return graph
+
+
+
+
+class Node(object):
+    """Node represents basic unit of graph"""
+    def __init__(self, data):
+        self.data = data
+
+    def __str__(self):
+        return 'Node({})'.format(self.data)
+    def __repr__(self):
+        return 'Node({})'.format(self.data)
+
+    def __eq__(self, other_node):
+        return self.data == other_node.data
+
+    def __ne__(self, other):
+        return not self.__eq__(other)
+
+    def __hash__(self):
+        return hash(self.data)
+
+class Edge(object):
+
+    def __init__(self, source, destination, weight):
+        self.source = source
+        self.destination = destination
+        self.weight = weight
+    def __str__(self):
+        return 'Edge(from {}, to {}, weight {})'.format(self.source, self.destination, self.weight)
+    def __repr__(self):
+        return 'Edge(from {}, to {}, weight {})'.format(self.source, self.destination, self.weight)
+
+    def __eq__(self, other_node):
+        return self.source == other_node.source and self.destination == other_node.destination and self.weight == other_node.weight
+    
+    def __ne__(self, other):
+        return not self.__eq__(other)
+
+    def __hash__(self):
+        return hash((self.source, self.destination, self.weight))
+
+
+class AdjacencyList(object):
+    """
+    AdjacencyList is one of the graph representation which uses adjacency list to
+    store nodes and edges
+    """
+    def __init__(self):
+        # adjacencyList should be a dictonary of node to edges
+        self.adjacency_list = {}
+        a = len(self.adjacency_list)
+     
+
+
+    def adjacent(self, node_1, node_2):
+    	if node_2 not in self.adjacency_list[node_1] :
+    		if node_1 not in self.adjacency_list[node_2]:
+    			return False
```

Comment:
This check looks like you are checking for the two way (which assumes the graph to be undirected graph)